### ISSUE TITLE
Multicast route slax

### ISSUE THREAD

**Issue by mmihir82** [Juniper/junoscriptorium#18]
Hello I need automatic using slax script, to run the command: "show multicast route extension" 
If the multicast group is in forwarding (multicast-route-forwarding-state), print me the group with forwarding-rate-kilobytes. There will be two entry of the same group. 
I'm totally new to this slax. 

`<rpc-reply xmlns:junos="http://xml.juniper.net/junos/14.1X53/junos">
    <multicast-route-information xmlns="http://xml.juniper.net/junos/14.1X53/junos-routing">
        <route-family>
            <multicast-instance>master</multicast-instance>
            <address-family>INET</address-family>
            <multicast-route junos:style="extensive">
                <multicast-group-address>232.0.1.0</multicast-group-address>
                <multicast-source-address>192.168.2.5</multicast-source-address>
                <multicast-source-prefix-length>32</multicast-source-prefix-length>
                <upstream-interface-name>ae1.0</upstream-interface-name>
                <downstream-interface-names>
                    <interface-name>irb.1</interface-name>
                </downstream-interface-names>
                <outgoing-interface-count>1</outgoing-interface-count>
                <session-name>Source specific multicast</session-name>
                <forwarding-rate-kilobytes>1465</forwarding-rate-kilobytes>
                <forwarding-rate-packets>982</forwarding-rate-packets>
                <forwarded-packet-count>224636156</forwarded-packet-count>
                <next-hop-id>131076</next-hop-id>
                <upstream-protocol>Upstream protocol: PIM</upstream-protocol>
                <multicast-route-state>Active</multicast-route-state>
                <multicast-route-forwarding-state>Forwarding</multicast-route-forwarding-state>
                <multicast-route-timeout>forever</multicast-route-timeout>
                <interface-mismatch-count>0</interface-mismatch-count>
                <multicast-route-uptime junos:seconds="169249">1d 23:00:49</multicast-route-uptime>
            </multicast-route>
            <multicast-route junos:style="extensive">
                <multicast-group-address>232.0.1.0</multicast-group-address>
                <multicast-source-address>10.10.10.5</multicast-source-address>
                <multicast-source-prefix-length>32</multicast-source-prefix-length>
                <upstream-interface-name>xe-0/0/1.0</upstream-interface-name>
                <downstream-interface-names>
                    <interface-name>ae1.0</interface-name>
                </downstream-interface-names>
                <outgoing-interface-count>1</outgoing-interface-count>
                <session-name>Source specific multicast</session-name>
                <forwarding-rate-kilobytes>0</forwarding-rate-kilobytes>
                <forwarding-rate-packets>587</forwarding-rate-packets>
                <forwarded-packet-count>193339235</forwarded-packet-count>
                <next-hop-id>131090</next-hop-id>
                <upstream-protocol>Upstream protocol: PIM</upstream-protocol>
                <multicast-route-state>Active</multicast-route-state>
                <multicast-route-forwarding-state>Forwarding</multicast-route-forwarding-state>
                <multicast-route-timeout>forever</multicast-route-timeout>
                <interface-mismatch-count>0</interface-mismatch-count>
                <multicast-route-uptime junos:seconds="252649">2d 22:10:49</multicast-route-uptime>
            </multicast-route>
        </route-family>
    </multicast-route-information>
</rpc-reply>`

**Comment by philshafer** [comment]
On Apr 17, 2017, at 10:25 PM, Mihir Mehta <notifications@github.com> wrote:
> Hello I need automatic using slax script, to run the command: "show multicast route extension"
> If the multicast group is in forwarding (multicast-route-forwarding-state), print me the group with forwarding-rate-kilobytes. There will be two entry of the same group.
> I'm totally new to this slax
> 


Given your XML, you’re wanting something like:

------------
version 1.2;
    
main <op-script-results> {
    var $rpc = <command> "show multicast route extension";
    var $res = jcs:invoke($rpc);
    mvar $hit = 0;
    for-each ($res/multicast-route-information/route-family) {
	var $mi = multicast-instance;
	var $fam = address-family;
	for-each (multicast-route) {
	    var $addr = multicast-group-address;
	    var $interfaces = downstream-interface-names/interface-name;
	    var $rate = forwarding-rate-kilobytes;

	    expr slax:printf("%12s%8s%12s%8s%s",
			    $mi, $fam, $addr, $rate, $interfaces);
	    set $hit = 1;
	}
    }
    if ($hit == 0) {
	expr slax:printf("no matches");
    }
}
---------------

Apologies that I don’t have a box to test this against, but this is the sketch of the code.

For information about slax, can I suggest the “Day One” book as a starting point:

https://www.juniper.net/us/en/training/jnbooks/day-one/automation-series/junos-automation-slax/ <https://www.juniper.net/us/en/training/jnbooks/day-one/automation-series/junos-automation-slax/>

There’s more “automation” material under:

https://www.juniper.net/us/en/training/jnbooks/day-one/ <https://www.juniper.net/us/en/training/jnbooks/day-one/>

Thanks,
 Phil

**Comment by mmihir82** [comment]
Hello Phil,
Thank you for providing the info. I will surely read the book, that you mention. It will be good start.
M

**Comment by mmihir82** [comment]
Hello Phil,
I had to modified the script to print the information out. Now I will need to add logic to print only one group, instead of printing twice. 
```
match /
{
    <op-script-results>
    {
        /* execute the 'show multicast route extension' command and store the result */
        var $rpc-query = { <get-multicast-route-information> { <extensive>; } }
        var $results = jcs:invoke($rpc-query);

        /* print a header */
        <output>"Multicast\tSource\t\tKilobytes\tStatus\t\tWrongInterface";

        /* search for the Multicast Route Group */
        <multicast-route-information> {
                /* print Multicast with WrongInterface counter */
                for-each($results/route-family/multicast-route) {
                  <output> multicast-group-address _ "\t" _ multicast-source-address _ "\t\t" _ forwarding-rate-kilobytes _ "\t" _ multicast-route-forwarding-state _ "\t\t" _ interface-mismatch-count;
                }
        }
    }
}
```
Here is the results:
```
Multicast	Source		Kilobytes	Status		WrongInterface		Upstream	DownStream
232.0.1.0	192.168.161.2		0	Forwarding		0		ae1.0		irb.1111
232.0.1.0	10.10.232.1		1122	Forwarding		0		xe-0/0/1.0	ae1.0
232.0.2.1	192.168.163.13		2154	Forwarding		0		ae1.0		irb.1111
232.0.2.1	10.10.232.1		1206	Forwarding		0		xe-0/0/8.0	ae1.0
232.0.2.2	192.168.163.13		0	Forwarding		0		ae1.0		irb.1111
```
Desired results:
```
Multicast		Source			Kilobytes		Status			WrongInterface			Upstream/DownStream	
232.0.1.0	192.168.161.2  ---- 10.10.232.1	0  ----  1122	Forwarding	0	ae1.0 ---- irb.1111
232.0.2.1	192.168.161.1  ---- 10.10.232.2	1112 ---- 1111	Forwarding	0	xe-0/0/1.0 ---- ae1.0

```

**Comment by philshafer** [comment]
Consider using the “%j1s” of printf, which only print the value when it changes:

   http://juniper.github.io/libslax/slax-manual.html#slaxprintf <http://juniper.github.io/libslax/slax-manual.html#slaxprintf>

Hmm… weak documentation, but it works like this:

            for-each (1 ... 4) {
                var $boom = {
                    if (position() == last()) {
                        expr " boom!!";
                    }
                }
                <one> slax:printf("testing %j1-4d... %4s... %s%s!",
                                  1, . + 1, "three", $boom);
            }

which makes the output:

    <one>testing 1   ...    2... three!</one>
    <one>testing     ...    3... three!</one>
    <one>testing     ...    4... three!</one>
    <one>testing     ...    5... three boom!!!</one>


Thanks,
 Phil

**Comment by mmihir82** [comment]
Thanks, Will try it out. Question: If I want to run a certain commands base on the match result and output into a file with date and time. How can i do that?

**Comment by philshafer** [comment]
On Apr 20, 2017, at 2:32 PM, Mihir Mehta <notifications@github.com> wrote:
>  If I want to run a certain commands base on the match result and output into a file with date and time. How can i do that?

Use the <redirect:write> to open and write files.  See the day-one book for details (page 302).

Use date:date() to get the current time.  Also in the day one book (page 100).

Thanks,
 Phil

**Comment by mmihir82** [comment]
awesome Phil.. thanks

**Comment by mmihir82** [comment]
Hello Phil,
I was able to write to the file. Issue running into, When I used 
```<saxon:output href=$filename method="text" append="yes">``` it prints fine, but how can I get the raw data, when running command like: "show pim join extensive | no-more". I need to collect data by running bunch of "show" commands and put it into the file. I was able to execute some shell commands, but when running "show commands" and append to file as "text" it remove bunch of other stuff and only print values... How can I get raw data save into the file? 
I used ``` <request-shell-execute> { <command> "ifconfig em0"; } ``` is there something I can run similar to excute?

Thanks
M

**Comment by philshafer** [comment]
On Apr 21, 2017, at 7:00 PM, Mihir Mehta <notifications@github.com> wrote:
>  When I used
> <saxon:output href=$filename method="text" append="yes"> it prints fine, but how can I get the raw data, when running command like: "show pim join extensive | no-more". I need to collect data by running bunch of "show" commands and put it into the file. I was able to execute some shell commands, but when running "show commands" and append to file as "text" it remove bunch of other stuff and only print values... How can I get raw data save into the file?

FWIW My preferences is to use <redirect:write>.

For writing data, use <copy-of> to avoid copying just the data:

    <text> $foo; /* this is only text */
    <copy> { copy-of $foo; } /* full copy */

but to get formatted text output (rather than XML content), you need to put the ‘format=“text”’ attribute on the RPC:

   <get-whatever format=“text”>;

Thanks,
 Phil

**Comment by mmihir82** [comment]
Thanks, Phil, will the copy-of will append to an existing file? How to reference that. (syntax)

**Comment by mmihir82** [comment]
Hello Phil, 
I tried using ```copy-of``` and ```redirect:write``` using both combination, I still got just the data. 
Not sure what I doing incorrect.. 
Here is what I got:
```
match /
{
    <op-script-results>
    {
        /* Output to the file */
        var $filename2 = "/var/tmp/collectcprod.log";

        <output> "show pim join extensive | no-more:";
        var $cmd1 = <command> "show pim join extensive | no-more";
        var $result1 = jcs:invoke($cmd1);

        <output> "show multicast route extensive | no-more:";
        var $cmd2 = <command> "show multicast route extensive | no-more";
        var $result2 = jcs:invoke($cmd2);

        <redirect:write href=$filename2 method="text" append="yes">
        {
            expr date:date-time() _ "\n";
            expr "==========================";
            copy-of $result1;
        }
        <xt:document href=$filename2 method="text" append="yes">
        {
            expr date:date-time() _ "\n";
            expr "==========================";
            copy-of $result2;
        }
    }
}
```
Results:
```
2017-04-22T06:28:57Z
==========================

PIM.master
INET

232.0.1.0
10.10.161.5


ae1.0
10.10.211.3

342
2w0d 04:15:28

017-04-22T06:28:57Z
==========================
master
INET

232.0.1.0
10.10.161.5
32
ae1.0

irb.1111

1
Source specific multicast
0
0
50513
131070
Upstream protocol: PIM
Active
Forwarding
forever
0
2w0d 04:15:30
```

**Comment by philshafer** [comment]
On Apr 22, 2017, at 2:05 AM, Mihir Mehta <notifications@github.com> wrote:
> Thanks, Phil, will the copy-of will append to an existing file? How to reference that. (syntax)
> 


copy-of copies a hierarchy of nodes.  Consider the following:

match / {
    copy {
        copy-of .;
    }
    as-string {
        expr .;
    }
}

<copy> will contain a copy of the input hierarchy, where as-string will contain a “string only” version of the input.

To append to a file, you need something like:

main <op-script-results> {
    var $rpc = <command format=“text”> “whatever”;
    var $res = jcs:invoke($rpc);
    <redirect:write append=“yes” file=“test.output”>
        copy-of $res;
    }
}

Thanks,
 Phil

**Comment by mmihir82** [comment]
Thank you that works.. didn't know I could add format inside the "command" 
I just started learning this maybe couple days back... I think I have learned a lot... thank you for prompt reply

**Comment by mmihir82** [comment]
Hello Phil,
When creating a file ``` var $filename2 = "/var/tmp/collectcprod.log"; ```
is there is anyway to create file with run-time date-time, instead of appending?
Thanks

**Comment by philshafer** [comment]
On Apr 25, 2017, at 12:20 AM, Mihir Mehta <notifications@github.com> wrote:
> When creating a file var $filename2 = "/var/tmp/collectcprod.log";
> is there is anyway to create file with run-time date-time, instead of appending?

Not sure I follow you, but if you want the name of the file to contain the date, just use the “concat” operator, which is “_”:

   var $filename = “/var/tmp/data-“ _ date:date() _ “.log”;

Thanks,
 Phil

**Comment by mmihir82** [comment]
Thanks Phil, will try it out.

**Comment by mmihir82** [comment]
Hey Phil,
Is there anyway to print proper date and time, here is the output I got from 
```var $filename1 = "/var/tmp/collectcproc_" _ date:date() _"_" _ date:time() _ ".log"; ```
Output: ``` collectcproc_2017-04-27_00%3A05%3A02.log```
```var $filename1 = "/var/tmp/collectcproc_" _ date:date-time() _ ".log";```
Output: ```collectcmd_2017-04-26T23%3A43%3A23Z.log```

**Comment by philshafer** [comment]
Strange, I don’t see this:

% cat ~/trash/test.slax
version 1.2;

main <top> {
    <date> date:date();
    <time> date:time();
    var $filename = "/var/tmp/collectcproc_" _ date:date() _ "_" _ date:time() _ ".log";
    <file> $filename;
}
% slaxproc -g -E ~/trash/test.slax
<?xml version="1.0"?>
<top>
  <date>2017-04-26-04:00</date>
  <time>23:18:30-04:00</time>
  <file>/var/tmp/collectcproc_2017-04-26-04:00_23:18:30-04:00.log</file>
</top>

Ah: “:” is not valid in filenames, since it can’t work on some operating systems (like windows), so <redirect:write> will replace them with the “%”-escaped versions.

Use str:replace() to change the “:”s into something else, like:

    var $fn = str:replace(date:date-time(), “:”, “.);

Thanks,
 Phil

**Comment by mmihir82** [comment]
Thanks, instead of using redirect:write, I used saxon:output and I am able
to print properly with time.

Thanks, sorry for asking question upon question...
is there a way to run the script every 5mins?

Thanks,
Mihir M

**Comment by philshafer** [comment]
On Apr 27, 2017, at 1:04 AM, Mihir Mehta <notifications@github.com> wrote:
> Thanks, sorry for asking question upon question…

No worries.

> is there a way to run the script every 5mins?

Take a look in the “day one” book at “event scripts”, specially generated events.

Thanks,
 Phil

**Comment by dssatya** [comment]
Adding to thread.

https://www.juniper.net/documentation/en_US/junos/topics/example/junos-script-automation-event-policy-generating-internal-event.html

thanks,
satya

**Comment by mmihir82** [comment]
Hello Phil,
With the result above: I collected the multicast group with the list: using append command:
```append $wronglist += $multigrp _ ", " ;```
```
232.0.1.129
232.0.1.132
232.0.1.133
232.0.1.142
232.0.1.146
```
How to loop thru each one in the "wronglist" run the command like: "clear pim join " _ $wronglist;
I have counter count on the side which counts how many wronglist are there, in this case 5. 
I try to do ``` for-each($wronglist) ``` gave me error ``` error: The 'select' expression does not evaluate to a node set.```
Don't how to put that in a ARRAY format instead of using append/
thanks

**Comment by philshafer** [comment]
Hard to follow you here, but I think you are trying to build a nodeset in an mvar, but you are missing the element.  If you append strings, you’ll end up with a single string (meaning no structure).   Take a look at this example:

Bock % cat trash/nodeset.slax
version 1.2;

var $input := {
    <value> "232.0.1.129";
    <value> "232.0.1.132";
    <value> "232.0.1.133";
    <value> "232.0.1.142";
    <value> "232.0.1.146";
}

main <top> {
    mvar $ns;

    for-each ($input/value) {
        append $ns += <my> .;
    }

    for-each ($ns/my) {
        <found> .;
    }
}
Bock % slaxproc -g -E trash/nodeset.slax
<?xml version="1.0"?>
<top>
  <found>232.0.1.129</found>
  <found>232.0.1.132</found>
  <found>232.0.1.133</found>
  <found>232.0.1.142</found>
  <found>232.0.1.146</found>
</top>

Thanks,
 Phil

**Comment by mmihir82** [comment]
thanks Phil, Only thing I had to change
``` for-each ($ns/my) { <found> .; } ``` and
